pyccel/ast/numpyext.py

```diff
@@ -1238,6 +1239,25 @@ class NumpyArctanh(NumpyUfuncUnary):
 
 #=======================================================================================
 
+
+class NumpySign(NumpyUfuncUnary):
+    """Represent a call to the sign function in the Numpy library"""
+    __slots__ = ('_x',)
+    name = 'sign'
+
+    def __init__(self, x):
+        super().__init__(x)
+        self._x = x
```

Comment:
`x` is a pyccel object so it should be saved in `_attribute_nodes`. Alternatively you could change the property to:
```python
@property
def x(self):
    return self._args[0]
```

Reply:
I don't see why the property should be changed as we have only one argument `x` in the `__init__` for now. Unless you mean `__init__` must be written like this:
```python
def __init__(self, *args):
    self._args = args
    super().__init__(self._args[0])
```

Reply:
The `_args` attribute is already set in the super class `PyccelInternalFunction`. I think that either this should be used:
```python
def __init__(self, x):
        super().__init__(x)

@property
def x(self):
    return self._args[0]
```
or if the argument is saved directly then it needs to be saved as an `_attribute_node` so that the `substitute` function is not broken. This can either be done by having `x` and `_args`:
```python
_attribute_nodes = ('_args', '_x')
def __init__(self, x):
        self._x = x
        super().__init__(x)

@property
def x(self):
    return self._x
```
or by just keeping `x`:
```python
_attribute_nodes = ('_x',)
def __init__(self, x):
        self._x = x
        super().__init__()

@property
def x(self):
    return self._x
```
In the second case you will need a specific `_print` function for `NumpySign` as `_print_NumpyUfuncBase` won't be able to use `.args`

Reply:
Good to know, I always wondered when something is printed with `_print_NumpyUfuncBase` or a specific `_print_` function for numpy functions. So the existence of `_attribute_nodes` defines the `_print` method.

Reply:
I didn't know about the `args` thingy, I think the class would be simpler now without having the property `x` or `__init__`.

Reply:
> Good to know, I always wondered when something is printed with `_print_NumpyUfuncBase` or a specific `_print_` function for numpy functions. So the existence of `_attribute_nodes` defines the `_print` method.

Not at all. The choice of which `_print` function is explained in the developer docs here : https://github.com/pyccel/pyccel/blob/master/developer_docs/overview.md#function-naming-conventionsfile-navigation

All pyccel classes which inherit from `pyccel.ast.basic.Basic` have the `_attribute_nodes` property, but sometimes it is inherited from a superclass. It is used in [`pyccel.ast.basic.Basic`](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/basic.py#L28-L402) for functions such as:
- checking if an object is inside another (e.g. is the variable `a` somewhere inside the function `f`)
- replacing an object with another (e.g. replacing an array `b` with an element `b[i]` in the unravelling)

Reply:
> I didn't know about the `args` thingy, I think the class would be simpler now without having the property `x` or `__init__`.

I agree the class could be reduced to:
```python
class NumpySign(NumpyUfuncUnary):
    """Represent a call to the sign function in the Numpy library"""
    __slots__ = ('_x',)
    name = 'sign'

    def _set_dtype_precision(self, x):
        self._dtype = x.dtype
        self._precision = default_precision[str_dtype(self._dtype)]
```

Reply:
fixed here https://github.com/pyccel/pyccel/pull/1169/commits/dc70a1011e6f1c78dded7b279c0ccaa74cc63573